ale_linters/kotlin/kotlinc.vim

```diff
@@ -0,0 +1,124 @@
+" Author: Francis Agyapong <francisgyapong2@gmail.com>
+" Description: A linter for the Kotlin programming language that uses kotlinc
+
+let g:ale_kotlin_kotlinc_options = get(g:, 'ale_kotlin_kotlinc_options', '')
+let g:ale_kotlin_kotlinc_enable_config = get(g:, 'ale_kotlin_kotlinc_enable_config', 0)
+let g:ale_kotlin_kotlinc_config_file = get(g:, 'ale_kotlin_kotlinc_config_file', '.ale_kotlinc_config')
+let g:ale_kotlin_kotlinc_classpath = get(g:, 'ale_kotlin_kotlinc_classpath', '')
+let g:ale_kotlin_kotlinc_sourcepath = get(g:, 'ale_kotlin_kotlinc_sourcepath', '')
+let g:ale_kotlin_kotlinc_use_module_file = get(g:, 'ale_kotlin_kotlinc_use_module_file', 0)
+let g:ale_kotlin_kotlinc_module_filename = get(g:, 'ale_kotlin_kotlinc_module_filename', 'module.xml')
+
+function! ale_linters#kotlin#kotlinc#GetCommand(buffer)
+    let l:kotlinc_opts = g:ale_kotlin_kotlinc_options
+    let l:command = 'kotlinc '
+
+    " If the config file is enabled and readable, source it
+    if g:ale_kotlin_kotlinc_enable_config
+        if filereadable(expand(g:ale_kotlin_kotlinc_config_file, 1))
+            execute 'source ' . fnameescape(expand(g:ale_kotlin_kotlinc_config_file, 1))
+        endif
+    endif
+
+    " If use module and module file is readable use that and return
+    if g:ale_kotlin_kotlinc_use_module_file
+        let l:module_filename = fnameescape(expand(g:ale_kotlin_kotlinc_module_filename, 1))
+
+        if filereadable(l:module_filename)
+            let l:kotlinc_opts .= ' -module ' . l:module_filename
+            let l:command .= 'kotlinc ' . l:kotlinc_opts
+
+            return l:command
+        endif
+    endif
+
+    " We only get here if not using module or the module file not readable
+    if g:ale_kotlin_kotlinc_classpath !=# ''
+        let l:kotlinc_opts .= ' -cp ' . g:ale_kotlin_kotlinc_classpath
+    endif
+
+    let l:fname = ''
+
+    if g:ale_kotlin_kotlinc_sourcepath !=# ''
+        let l:fname .= expand(g:ale_kotlin_kotlinc_sourcepath, 1) . ' '
+    endif
+    let l:fname .= shellescape(expand('#' . a:buffer . ':p'))
+    let l:command .= l:kotlinc_opts . ' ' . l:fname
```

Comment:
Will this compile on the file on disk, or on the text buffer in Vim? If it's the former, then you need to use the stdin_wrapper command so the Vim buffer is copied to a temporary file. Otherwise, you won't be able to check warnings or errors until the file is saved.

If this compiler is anything like `javac`, then you probably won't be able to implement this until after the Windows stdin-wrapper script has been written to save files in temporary directories, and both stdin-wrapper scripts copy the basename of files to the temporary directories verbatim. See issues #40 and #117.